core/src/main/java/io/questdb/cairo/ColumnVersionPurgeJob.java

```diff
@@ -0,0 +1,349 @@
+        this.sqlCompiler.compile(
+                "CREATE TABLE IF NOT EXISTS \"" + tableName + "\" (" +
+                        "ts timestamp, " + // 0
+                        "table_name symbol, " + // 1
+                        "column_name symbol, " + // 2
+                        "table_id int, " + // 3
+                        "truncate_version int, " + // 4
+                        "columnType int, " + // 5
+                        "table_partition_by int, " + // 6
+                        "updated_txn long, " + // 7
+                        "column_version long, " + // 8
+                        "partition_timestamp timestamp, " + // 9
+                        "partition_name_txn long," + // 10
+                        "completed timestamp" + // 11
+                        ") timestamp(ts) partition by MONTH",
+                sqlExecutionContext
+        );
+        this.writer = engine.getWriter(AllowAllCairoSecurityContext.INSTANCE, getLogTableName(), "QuestDB system");
+        this.cleanupExecution = new ColumnVersionPurgeExecution(configuration, this.writer, "completed");
+    }
+
+    @Override
+    public void close() {
+        this.writer = Misc.free(this.writer);
+        this.sqlCompiler = Misc.free(sqlCompiler);
+        this.sqlExecutionContext = Misc.free(sqlExecutionContext);
+        this.cleanupExecution = Misc.free(cleanupExecution);
+    }
+
+    public String getLogTableName() {
+        return tableName;
+    }
+
+    private static int compareHouseKeepingTasks(ColumnVersionPurgeTaskRun task1, ColumnVersionPurgeTaskRun task2) {
+        return Long.compare(task1.nextRunTimestamp, task2.nextRunTimestamp);
+    }
+
+    private void calculateNextTimestamp(ColumnVersionPurgeTaskRun task, long currentTime) {
+        task.waitToRun = Math.min(maxWaitCapMicro, (long) (task.waitToRun * exponentialWaitMultiplier));
+        task.nextRunTimestamp = currentTime + task.waitToRun;
+    }
+
+    private boolean cleanup() {
+        boolean useful = false;
+        final long now = clock.getTicks() + 1;
+        while (houseKeepingRunQueue.size() > 0) {
+            ColumnVersionPurgeTaskRun next = houseKeepingRunQueue.peek();
+            if (next.nextRunTimestamp < now) {
+                houseKeepingRunQueue.poll();
+                useful = true;
+                if (!cleanupExecution.tryCleanup(next)) {
+                    // Re-queue
+                    calculateNextTimestamp(next, now);
+                    houseKeepingRunQueue.add(next);
+                }
+            } else {
+                // All reruns are in the future.
+                return useful;
+            }
+        }
+        return useful;
+    }
+
+    private void commit() {
+        try {
+            writer.commit();
+        } catch (Throwable th) {
+            LOG.error().$("error saving to column version house keeping log, cannot commit [table=").$(tableName).$(", error=").$(th).I$();
+            writer.rollback();
```

Comment:
free will perform rollback, do not call rollback explicitly

Reply:
done